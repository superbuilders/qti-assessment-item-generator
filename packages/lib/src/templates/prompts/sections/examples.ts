import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
