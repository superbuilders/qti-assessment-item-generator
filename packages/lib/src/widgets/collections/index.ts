import { allWidgetsCollection } from "@/widgets/collections/all"
import { fourthGradeMathCollection } from "@/widgets/collections/fourth-grade-math"
import { mathCoreCollection } from "@/widgets/collections/math-core"
import { scienceCollection } from "@/widgets/collections/science"
import { simpleVisualCollection } from "@/widgets/collections/simple-visual"
import { teksMath4Collection } from "@/widgets/collections/teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "@/widgets/collections/types"
