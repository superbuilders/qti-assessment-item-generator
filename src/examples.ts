import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
