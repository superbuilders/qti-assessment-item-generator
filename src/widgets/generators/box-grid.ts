import * as errors from "@superbuilders/errors"
import * as logger from "@superbuilders/slog"
import { z } from "zod"
import { CanvasImpl } from "../utils/canvas-impl"
import { PADDING } from "../utils/constants"
import { CSS_COLOR_PATTERN } from "../utils/css-color"
import { MATHML_INNER_PATTERN } from "../utils/mathml"
import { numberContentToInnerMathML } from "../utils/number-to-mathml"
import { createHeightSchema, createWidthSchema } from "../utils/schemas"
import { theme } from "../utils/theme"
import type { WidgetGenerator } from "../types"

// Defines the content and styling for a single cell in the grid.
const BoxGridCellSchema = z
	.object({
		content: z
			.discriminatedUnion("type", [
				// text-only removed: we render numbers/math as MathML for consistent typography
				z
					.object({
						type: z.literal("math").describe("Identifies this as mathematical content"),
						mathml: z
							.string()
							.regex(MATHML_INNER_PATTERN, "invalid mathml snippet; must be inner MathML without outer <math> wrapper")
							.describe("Inner MathML markup (no outer <math> element)")
					})
					.strict()
					.describe("Mathematical content represented in MathML format"),
				z
					.object({
						type: z.literal("whole").describe("Whole number content"),
						value: z.number().int().describe("Integer magnitude (non-directional)"),
						sign: z.enum(["+", "-"]).describe("Sign of the number")
					})
					.strict()
					.describe("Whole number content with explicit sign"),
				z
					.object({
						type: z.literal("fraction").describe("Fractional number content"),
						numerator: z.number().int().min(0).describe("Numerator (non-negative)"),
						denominator: z.number().int().positive().describe("Denominator (positive)"),
						sign: z.enum(["+", "-"]).describe("Sign of the fraction")
					})
					.strict()
					.describe("Fractional number content with explicit sign"),
				z
					.object({
						type: z.literal("mixed").describe("Mixed number content like 1 1/4"),
						whole: z.number().int().min(0).describe("Whole part (non-negative)"),
						numerator: z.number().int().min(0).describe("Numerator of the fractional part"),
						denominator: z.number().int().positive().describe("Denominator of the fractional part (positive)"),
						sign: z.enum(["+", "-"]).describe("Sign of the mixed number")
					})
					.strict()
					.describe("Mixed number content with explicit sign")
			])
			.describe("Inline content for the cell: math (MathML) or number (whole/fraction/mixed)"),
		backgroundColor: z
			.string()
			.regex(CSS_COLOR_PATTERN, "invalid css color; use hex (#RGB, #RRGGBB, #RRGGBBAA)")
			.nullable()
			.transform((val) => (val === "null" || val === "NULL" || val === "" ? null : val))
			.describe(
				"Hex-only color for the cell background (e.g., '#FFE5B4', '#1E90FF', '#FF00004D' for ~30% alpha, null). Null means no background color (transparent)."
			)
	})
	.strict()

// The main Zod schema for the boxGrid function.
export const BoxGridPropsSchema = z
	.object({
		type: z
			.literal("boxGrid")
			.describe("Identifies this as a box grid widget for displaying tabular data with optional cell highlighting."),
		width: createWidthSchema(),
		height: createHeightSchema(),
		data: z
			.array(z.array(BoxGridCellSchema))
			.describe(
				"2D array where data[row][col] represents the cell content. First row often contains headers. All rows should have the same number of columns for proper alignment."
			),
		showGridLines: z
			.boolean()
			.describe(
				"Whether to show borders between cells. True creates a traditional table look. False creates a borderless layout, useful for highlighting patterns."
			)
	})
	.strict()
	.describe(
		"Creates a grid/table structure for displaying data, patterns, or mathematical arrays. Supports individual cell highlighting for emphasis. Useful for multiplication tables, data organization, and pattern recognition."
	)

export type BoxGridProps = z.infer<typeof BoxGridPropsSchema>

/**
 * Generates an SVG diagram of a grid of cells, with each cell capable of
 * displaying data and having a custom background color for highlighting.
 */
export const generateBoxGrid: WidgetGenerator<typeof BoxGridPropsSchema> = async (props) => {
	const { width, height, data, showGridLines } = props

	const numRows = data.length
	if (numRows === 0) {
		logger.error("box grid has no rows", { width, height })
		throw errors.new("box grid: no rows")
	}
	const firstRow = data[0]
	const expectedNumCols = firstRow ? firstRow.length : 0
	if (expectedNumCols === 0) {
		logger.error("box grid first row has no columns", { width, height })
		throw errors.new("box grid: no columns")
	}
	for (let r = 0; r < numRows; r++) {
		const row = data[r]
		const rowCols = row ? row.length : -1
		if (rowCols !== expectedNumCols) {
			logger.error("box grid rows have inconsistent column counts", {
				rowIndex: r,
				expected: expectedNumCols,
				actual: rowCols
			})
			throw errors.new("box grid: ragged rows")
		}
	}
	const numCols = expectedNumCols

	const cellWidth = width / numCols
	const cellHeight = height / numRows

	const canvas = new CanvasImpl({
		chartArea: { left: 0, top: 0, width, height },
		fontPxDefault: 14,
		lineHeightDefault: 1.2
	})

	canvas.addStyle(`
        .cell-text {
            font-size: 14px;
            text-anchor: middle;
            dominant-baseline: middle;
        }
    `)

	// Loop through data to draw backgrounds and text content
	for (let r = 0; r < numRows; r++) {
		for (let c = 0; c < numCols; c++) {
			const cell = data[r]?.[c]
			if (!cell) continue

			const x = c * cellWidth
			const y = r * cellHeight

			// Draw background rectangle for highlighting, if specified
			if (cell.backgroundColor && cell.backgroundColor !== "") {
				canvas.drawRect(x, y, cellWidth, cellHeight, {
					fill: cell.backgroundColor
				})
			}

			// Draw content: for math and numbers, render as MathML via XHTML foreignObject for typographic quality
			if (
				cell.content.type === "math" ||
				cell.content.type === "whole" ||
				cell.content.type === "fraction" ||
				cell.content.type === "mixed"
			) {
				const inner = cell.content.type === "math" ? cell.content.mathml : numberContentToInnerMathML(cell.content)

				const fontPx = 14
				// Increase font size for better legibility in grid cells
				const xhtml = `<!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:${theme.font.family.sans};color:${theme.colors.text};"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:${fontPx * 1.25}px;">${inner}</math></div>`

				canvas.drawForeignObject({
					x: x,
					y: y,
					width: cellWidth,
					height: cellHeight,
					content: xhtml
				})
			}
		}
	}

	// Loop again to draw grid lines on top of all content
	if (showGridLines) {
		for (let r = 0; r < numRows; r++) {
			for (let c = 0; c < numCols; c++) {
				const x = c * cellWidth
				const y = r * cellHeight
				canvas.drawRect(x, y, cellWidth, cellHeight, {
					fill: "none",
					stroke: theme.colors.border,
					strokeWidth: theme.stroke.width.base
				})
			}
		}
	}

	// NEW: Finalize the canvas and construct the root SVG element
	const { svgBody, vbMinX, vbMinY, width: finalWidth, height: finalHeight } = canvas.finalize(PADDING)

	return `<svg width="${finalWidth}" height="${finalHeight}" viewBox="${vbMinX} ${vbMinY} ${finalWidth} ${finalHeight}" xmlns="http://www.w3.org/2000/svg" font-family="${theme.font.family.sans}">${svgBody}</svg>`
}
