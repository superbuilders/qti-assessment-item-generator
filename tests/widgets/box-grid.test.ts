import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import { boxGridExamples } from "../../examples/box-grid"
import type { Widget } from "../../src/widgets/registry"
import { generateWidget } from "../../src/widgets/widget-generator"

describe("Widget: box-grid", () => {
	const examples = boxGridExamples as unknown as Widget[]

	examples.forEach((props, index) => {
		test(`should produce consistent output for example #${index + 1}`, async () => {
			const result = await errors.try(generateWidget(props))
			if (result.error) throw result.error
			expect(result.data).toMatchSnapshot()
		})
	})
})


