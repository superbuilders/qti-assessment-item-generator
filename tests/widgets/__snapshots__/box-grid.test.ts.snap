// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget: box-grid should produce consistent output for example #1 1`] = `
"<svg width="342" height="342" viewBox="-21 -21 342 342" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif"><defs><style>
        .cell-text {
            font-size: 14px;
            text-anchor: middle;
            dominant-baseline: middle;
        }
    </style></defs><rect x="0" y="0" width="75" height="100" fill="#FFE5B4"/><foreignObject x="0" y="0" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>1</mn></math></div></foreignObject><foreignObject x="75" y="0" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mfrac><mn>1</mn><mn>2</mn></mfrac></math></div></foreignObject><rect x="150" y="0" width="75" height="100" fill="#B4E5FF"/><foreignObject x="150" y="0" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>1</mn><mspace width="0.15em"/><mfrac><mn>1</mn><mn>4</mn></mfrac></math></div></foreignObject><foreignObject x="225" y="0" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>4</mn></math></div></foreignObject><foreignObject x="0" y="100" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>5</mn></math></div></foreignObject><rect x="75" y="100" width="75" height="100" fill="#FFB4E5"/><foreignObject x="75" y="100" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mo>-</mo><mn>2</mn><mspace width="0.15em"/><mfrac><mn>3</mn><mn>5</mn></mfrac></math></div></foreignObject><foreignObject x="150" y="100" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mfrac><mn>3</mn><mn>4</mn></mfrac></math></div></foreignObject><rect x="225" y="100" width="75" height="100" fill="#B4FFE5"/><foreignObject x="225" y="100" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>8</mn></math></div></foreignObject><foreignObject x="0" y="200" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>0</mn></math></div></foreignObject><rect x="75" y="200" width="75" height="100" fill="#E5B4FF"/><foreignObject x="75" y="200" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>3</mn></math></div></foreignObject><foreignObject x="150" y="200" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mo>-</mo><mfrac><mn>7</mn><mn>8</mn></mfrac></math></div></foreignObject><rect x="225" y="200" width="75" height="100" fill="#FFE5B4"/><foreignObject x="225" y="200" width="75" height="100"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>12</mn></math></div></foreignObject><rect x="0" y="0" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="75" y="0" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="150" y="0" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="225" y="0" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="0" y="100" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="75" y="100" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="150" y="100" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="225" y="100" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="0" y="200" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="75" y="200" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="150" y="200" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="225" y="200" width="75" height="100" fill="none" stroke="#e0e0e0" stroke-width="1.5"/></svg>"
`;

exports[`Widget: box-grid should produce consistent output for example #2 1`] = `
"<svg width="342" height="342" viewBox="-21 -21 342 342" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif"><defs><style>
        .cell-text {
            font-size: 14px;
            text-anchor: middle;
            dominant-baseline: middle;
        }
    </style></defs><rect x="0" y="0" width="100" height="150" fill="#4285F4"/><foreignObject x="0" y="0" width="100" height="150"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>2</mn><mo>+</mo><mn>3</mn></math></div></foreignObject><foreignObject x="100" y="0" width="100" height="150"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>0</mn></math></div></foreignObject><rect x="200" y="0" width="100" height="150" fill="#34A853"/><foreignObject x="200" y="0" width="100" height="150"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>5</mn></math></div></foreignObject><rect x="0" y="150" width="100" height="150" fill="#FF6B6B"/><foreignObject x="0" y="150" width="100" height="150"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>4</mn><mo>Ã—</mo><mn>6</mn></math></div></foreignObject><foreignObject x="100" y="150" width="100" height="150"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>0</mn></math></div></foreignObject><rect x="200" y="150" width="100" height="150" fill="#FFD93D"/><foreignObject x="200" y="150" width="100" height="150"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:center;justify-content:center;width:100%;height:100%;font-family:sans-serif;color:#333333;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.5px;"><mn>24</mn></math></div></foreignObject><rect x="0" y="0" width="100" height="150" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="100" y="0" width="100" height="150" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="200" y="0" width="100" height="150" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="0" y="150" width="100" height="150" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="100" y="150" width="100" height="150" fill="none" stroke="#e0e0e0" stroke-width="1.5"/><rect x="200" y="150" width="100" height="150" fill="none" stroke="#e0e0e0" stroke-width="1.5"/></svg>"
`;
